File: src/main/java/pl/yalgrin/playnite/simplesync/dto/GameDiffDTO.java
package pl.yalgrin.playnite.simplesync.dto;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AccessLevel;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import pl.yalgrin.playnite.simplesync.utils.ToStringUtils;

import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.util.List;

import static pl.yalgrin.playnite.simplesync.dto.GameDiffDTO.Fields.*;

@EqualsAndHashCode(callSuper = true)
@Data
@NoArgsConstructor
@SuperBuilder
public class GameDiffDTO extends AbstractDiffDTO {

    @JsonProperty(DESCRIPTION)
    private String description;

    @JsonProperty(NOTES)
    private String notes;

    @JsonProperty(GENRES)
    private List<GenreDTO> genres;

    @JsonProperty(HIDDEN)
    private boolean hidden;

    @JsonProperty(FAVORITE)
    private boolean favorite;

    @JsonProperty(LAST_ACTIVITY)
    private ZonedDateTime lastActivity;

    @JsonProperty(SORTING_NAME)
    private String sortingName;

    @JsonProperty(GAME_ID)
    private String gameId;

    @JsonProperty(PLUGIN_ID)
    private String pluginId;

    @JsonProperty(INCLUDE_LIBRARY_PLUGIN_ACTION)
    private boolean includeLibraryPluginAction;

    @JsonProperty(PLATFORMS)
    private List<PlatformDTO> platforms;

    @JsonProperty(PUBLISHERS)
    private List<CompanyDTO> publishers;

    @JsonProperty(DEVELOPERS)
    private List<CompanyDTO> developers;

    @JsonProperty(RELEASE_DATE)
    private LocalDateTime releaseDate;

    @JsonProperty(CATEGORIES)
    private List<CategoryDTO> categories;

    @JsonProperty(TAGS)
    private List<TagDTO> tags;

    @JsonProperty(FEATURES)
    private List<FeatureDTO> features;

    @JsonProperty(LINKS)
    private List<LinkDTO> links;

    @JsonProperty(PLAYTIME)
    private long playtime;

    @JsonProperty(PLAYTIME_DIFF)
    private Long playtimeDiff;

    @JsonProperty(ADDED)
    private ZonedDateTime added;

    @JsonProperty(MODIFIED)
    private ZonedDateTime modified;

    @JsonProperty(PLAY_COUNT)
    private long playCount;

    @JsonProperty(PLAY_COUNT_DIFF)
    private Long playCountDiff;

    @JsonProperty(INSTALL_SIZE)
    private Long installSize;

    @JsonProperty(LAST_SIZE_SCAN_DATE)
    private ZonedDateTime lastSizeScanDate;

    @JsonProperty(SERIES)
    private List<SeriesDTO> series;

    @JsonProperty(VERSION)
    private String version;

    @JsonProperty(AGE_RATINGS)
    private List<AgeRatingDTO> ageRatings;

    @JsonProperty(REGIONS)
    private List<RegionDTO> regions;

    @JsonProperty(SOURCE)
    private SourceDTO source;

    @JsonProperty(COMPLETION_STATUS)
    private CompletionStatusDTO completionStatus;

    @JsonProperty(USER_SCORE)
    private Integer userScore;

    @JsonProperty(CRITIC_SCORE)
    private Integer criticScore;

    @JsonProperty(COMMUNITY_SCORE)
    private Integer communityScore;

    @JsonProperty(MANUAL)
    private String manual;

    @Override
    public String toString() {
        return ToStringUtils.createBuilder(this)
                .append("id", id)
                .append("name", name)
                .append("removed", removed)
                .toString();
    }

    @NoArgsConstructor(access = AccessLevel.PROTECTED)
    public static class Fields extends AbstractObjectDTO.Fields {
        public static final String DESCRIPTION = "Description";
        public static final String NOTES = "Notes";
        public static final String GENRES = "Genres";
        public static final String HIDDEN = "Hidden";
        public static final String FAVORITE = "Favorite";
        public static final String LAST_ACTIVITY = "LastActivity";
        public static final String SORTING_NAME = "SortingName";
        public static final String GAME_ID = "GameId";
        public static final String PLUGIN_ID = "PluginId";
        public static final String INCLUDE_LIBRARY_PLUGIN_ACTION = "IncludeLibraryPluginAction";
        public static final String PLATFORMS = "Platforms";
        public static final String PUBLISHERS = "Publishers";
        public static final String DEVELOPERS = "Developers";
        public static final String RELEASE_DATE = "ReleaseDate";
        public static final String CATEGORIES = "Categories";
        public static final String TAGS = "Tags";
        public static final String FEATURES = "Features";
        public static final String LINKS = "Links";
        public static final String PLAYTIME = "Playtime";
        public static final String PLAYTIME_DIFF = "PlaytimeDiff";
        public static final String ADDED = "Added";
        public static final String MODIFIED = "Modified";
        public static final String PLAY_COUNT = "PlayCount";
        public static final String PLAY_COUNT_DIFF = "PlayCountDiff";
        public static final String INSTALL_SIZE = "InstallSize";
        public static final String LAST_SIZE_SCAN_DATE = "LastSizeScanDate";
        public static final String SERIES = "Series";
        public static final String VERSION = "Version";
        public static final String AGE_RATINGS = "AgeRatings";
        public static final String REGIONS = "Regions";
        public static final String SOURCE = "Source";
        public static final String COMPLETION_STATUS = "CompletionStatus";
        public static final String USER_SCORE = "UserScore";
        public static final String CRITIC_SCORE = "CriticScore";
        public static final String COMMUNITY_SCORE = "CommunityScore";
        public static final String MANUAL = "Manual";
    }
}

File: src/main/kotlin/pl/yalgrin/playnite/simplesync/library/dto/GameDiffDTO.kt
package pl.yalgrin.playnite.simplesync.library.dto

import com.fasterxml.jackson.annotation.JsonProperty
import pl.yalgrin.playnite.simplesync.common.utils.ToStringUtils
import java.io.Serial
import java.time.LocalDateTime
import java.time.ZonedDateTime

data class GameDiffDTO(
    override var id: String? = null,
    override var name: String? = null,
    override var baseObjectId: Long? = null,
    override var changedFields: List<String> = emptyList(),
    @param:JsonProperty("isRemoved")
    @get:JsonProperty("isRemoved")
    @field:JsonProperty("isRemoved")
    override var isRemoved: Boolean = false,
    var description: String? = null,
    var notes: String? = null,
    var genres: List<GenreDTO> = emptyList(),
    @param:JsonProperty("isHidden")
    @get:JsonProperty("isHidden")
    @field:JsonProperty("isHidden")
    var isHidden: Boolean = false,
    @param:JsonProperty("isFavorite")
    @get:JsonProperty("isFavorite")
    @field:JsonProperty("isFavorite")
    var isFavorite: Boolean = false,
    var lastActivity: ZonedDateTime? = null,
    var sortingName: String? = null,
    var gameId: String? = null,
    var pluginId: String? = null,
    var platforms: List<PlatformDTO> = emptyList(),
    var publishers: List<CompanyDTO> = emptyList(),
    var developers: List<CompanyDTO> = emptyList(),
    var releaseDate: LocalDateTime? = null,
    var categories: List<CategoryDTO> = emptyList(),
    var tags: List<TagDTO> = emptyList(),
    var features: List<FeatureDTO> = emptyList(),
    var links: List<LinkDTO> = emptyList(),
    var playtime: Long = 0,
    var playtimeDiff: Long? = null,
    var added: ZonedDateTime? = null,
    var modified: ZonedDateTime? = null,
    var playCount: Long = 0,
    var playCountDiff: Long? = null,
    var installSize: Long? = null,
    var lastSizeScanDate: ZonedDateTime? = null,
    var series: List<SeriesDTO> = emptyList(),
    var version: String? = null,
    var ageRatings: List<AgeRatingDTO> = emptyList(),
    var regions: List<RegionDTO> = emptyList(),
    var source: SourceDTO? = null,
    var completionStatus: CompletionStatusDTO? = null,
    var userScore: Int? = null,
    var criticScore: Int? = null,
    var communityScore: Int? = null,
    var manual: String? = null
) : LibraryObjectDiffDTO {

    override fun toString(): String {
        return ToStringUtils.createBuilder(this)
            .append("id", id)
            .append("name", name)
            .append("isRemoved", isRemoved)
            .toString()
    }

    companion object {
        @Serial
        private const val serialVersionUID = -5504426661125673615L
    }
}

class GameDiffFields {
    companion object {
        const val ID: String = "Id"
        const val NAME: String = "Name"
        const val REMOVED: String = "Removed"
        const val DESCRIPTION: String = "Description"
        const val NOTES: String = "Notes"
        const val GENRES: String = "Genres"
        const val HIDDEN: String = "Hidden"
        const val FAVORITE: String = "Favorite"
        const val LAST_ACTIVITY: String = "LastActivity"
        const val SORTING_NAME: String = "SortingName"
        const val GAME_ID: String = "GameId"
        const val PLUGIN_ID: String = "PluginId"
        const val PLATFORMS: String = "Platforms"
        const val PUBLISHERS: String = "Publishers"
        const val DEVELOPERS: String = "Developers"
        const val RELEASE_DATE: String = "ReleaseDate"
        const val CATEGORIES: String = "Categories"
        const val TAGS: String = "Tags"
        const val FEATURES: String = "Features"
        const val LINKS: String = "Links"
        const val PLAYTIME: String = "Playtime"
        const val PLAYTIME_DIFF: String = "PlaytimeDiff"
        const val ADDED: String = "Added"
        const val MODIFIED: String = "Modified"
        const val PLAY_COUNT: String = "PlayCount"
        const val PLAY_COUNT_DIFF: String = "PlayCountDiff"
        const val INSTALL_SIZE: String = "InstallSize"
        const val LAST_SIZE_SCAN_DATE: String = "LastSizeScanDate"
        const val SERIES: String = "Series"
        const val VERSION: String = "Version"
        const val AGE_RATINGS: String = "AgeRatings"
        const val REGIONS: String = "Regions"
        const val SOURCE: String = "Source"
        const val COMPLETION_STATUS: String = "CompletionStatus"
        const val USER_SCORE: String = "UserScore"
        const val CRITIC_SCORE: String = "CriticScore"
        const val COMMUNITY_SCORE: String = "CommunityScore"
        const val MANUAL: String = "Manual"
    }
}